src/JestExt.ts

```diff
@@ -445,79 +445,121 @@ export class JestExt {
   /**
    * Primitive way to resolve path to jest.js
    */
-  private resolvePathToJestBin() {
+  private resolveTestProgram(): { program: string; args: string[]; isCreateReactApp: boolean } {
+    // Let's start with the given command
     let jest = this.workspace.pathToJest
+    let basename = path.basename(jest)
+    let isCreateReactApp = false
+
+    if (basename.indexOf('npm test') === 0 || basename.indexOf('npm.cmd test') === 0) {
+      // We can't always debug running npm test, therefore we have to dig deeper.
+      try {
+        const packagePath = path.join(vscode.workspace.rootPath, 'package.json')
+        const packageJSON = JSON.parse(readFileSync(packagePath, 'utf8'))
+        if (!packageJSON || !packageJSON.scripts || !packageJSON.scripts.test) {
+          throw 'invalid package.json'
+        }
+        // If the initial command contained additional parameters (separated by `--`),
+        // add them to the test command as if npm would do it.
+        const args = basename.split('0').slice(3)
+        basename = packageJSON.scripts.test + ' ' + args.join(' ')
+        // If the test script given in package.json hasn't the form of a relative path,
+        // we assume that it resides in `node_modules/.bin`.
+        if (basename.substr(0, 1) !== '.') {
+          jest = 'node_modules/.bin/' + basename
+        }
+        isCreateReactApp = testCommandIsCreateReactApp(packageJSON.scripts.test)
+      } catch {
+        vscode.window.showErrorMessage("package.json couldn't be read!")
+        return undefined
+      }
+    }
+
     if (!path.isAbsolute(jest)) {
       jest = path.join(vscode.workspace.rootPath, jest)
     }
 
-    const basename = path.basename(jest)
-    switch (basename) {
-      case 'jest.js': {
-        return jest
-      }
+    // We hope for the best that the filename doesn't contain any spaces and separate
+    // the basename from the arguments.
+    const args = basename.split(' ')

```

Comment:
Adding an external dependency would probably be the best way for a correct separation of arguments (including quotes)?

Reply:
(For example https://www.npmjs.com/package/string-argv)

Reply:
The root cause of this problem is `jest.pathToJest` is ambiguous as a string, and hard to parse into the parameters we need to provide to `child_process.spawn`. Why not remove the problem by asking for the type signature we need: `{ cmd: String, args: String[] }`.

You're going to see this again soon as a suggestion to fix the regression we've introduced on Windows.

Reply:
That would work for the settings, but not, if one sets it to npm test, in which case we would still end with the command string from package.json.

Which regression do you mean?

Reply:
I'm not sure I understand the issue. Is the problem that a user who has requested that we run the "npm test" script to start Jest is forcing us to find the path to the JavaScript file?

Would the args be passed to `node` if we spawned that command?